File: shared/src/windows/event.rs
use crate::windows::handle::{self, handle_wrapper};
use std::io;
use thiserror::Error;
use winapi::{
    shared::{
        minwindef::{FALSE, TRUE},
        ntdef::NULL,
        winerror::WAIT_TIMEOUT,
    },
    um::{
        minwinbase::SECURITY_ATTRIBUTES,
        synchapi::{CreateEventA, ResetEvent, SetEvent, WaitForSingleObject},
        winbase::{WAIT_FAILED, WAIT_OBJECT_0},
    },
};

handle_wrapper!(ManualResetEvent);

impl ManualResetEvent {
    pub fn new() -> Result<Self, NewError> {
        let security_attributes = SECURITY_ATTRIBUTES {
            #[expect(clippy::cast_possible_truncation)]
            nLength: size_of::<SECURITY_ATTRIBUTES>() as u32,
            lpSecurityDescriptor: NULL,
            bInheritHandle: FALSE,
        };

        unsafe {
            let handle = CreateEventA(
                std::ptr::addr_of!(security_attributes).cast_mut(),
                TRUE,
                FALSE,
                NULL.cast(),
            );
            if handle == NULL {
                return Err(io::Error::last_os_error().into());
            }
            Ok(Self::from_raw_handle(handle))
        }
    }

    pub fn get(&self) -> Result<bool, GetError> {
        unsafe {
            match WaitForSingleObject(self.handle.as_raw(), 0) {
                WAIT_OBJECT_0 => Ok(true),
                WAIT_TIMEOUT => Ok(false),
                WAIT_FAILED => Err(io::Error::last_os_error().into()),
                _ => unreachable!(),
            }
        }
    }

    pub async fn wait(&self) -> Result<(), WaitError> {
        self.handle.wait().await?;
        Ok(())
    }

    pub fn set(&mut self) -> Result<(), SetError> {
        unsafe {
            if SetEvent(self.handle.as_raw()) == 0 {
                return Err(io::Error::last_os_error().into());
            }
            Ok(())
        }
    }

    pub fn reset(&mut self) -> Result<(), ResetError> {
        unsafe {
            if ResetEvent(self.handle.as_raw()) == 0 {
                return Err(io::Error::last_os_error().into());
            }
            Ok(())
        }
    }
}

#[derive(Debug, Error)]
#[error("failed to create event")]
pub struct NewError(#[from] io::Error);

#[derive(Debug, Error)]
#[error("failed to clone event")]
pub struct CloneError(#[from] handle::CloneError);

#[derive(Debug, Error)]
#[error("failed to get event state")]
pub struct GetError(#[from] io::Error);

#[derive(Debug, Error)]
#[error("failed to wait for event")]
pub enum WaitError {
    HandleWait(#[from] handle::WaitError),
}

#[derive(Debug, Error)]
#[error("failed to set event to signaled state")]
pub struct SetError(#[from] io::Error);

#[derive(Debug, Error)]
#[error("failed to reset event to non-signaled state")]
pub struct ResetError(#[from] io::Error);

File: winter/tests/programs/src/saved_state/event.c
#include <stdbool.h>
#include <stdio.h>
#include <windows.h>

#include "assert.h"

bool event_get_state(HANDLE event)
{
    switch (WaitForSingleObject(event, 0))
    {
    case WAIT_OBJECT_0:
        return true;

    case WAIT_TIMEOUT:
        return false;

    default:
        ASSERT_WINAPI(false)
    }
}

void event_set_state(HANDLE event, bool state)
{
    BOOL result;
    if (state)
    {
        result = SetEvent(event);
    }
    else
    {
        result = ResetEvent(event);
    }
    ASSERT_WINAPI(result)
}

int main()
{
    HANDLE *i_bits = malloc(sizeof(HANDLE) * 3);
    for (int j = 0; j < 3; j++)
    {
        i_bits[j] = CreateEvent(NULL, j % 2 ? TRUE : FALSE, FALSE, NULL);
        ASSERT_WINAPI(i_bits[j]);
    }

    while (true)
    {
        int i = 0;
        for (int j = 0; j < 3; j++)
        {
            i |= event_get_state(i_bits[j]) << j;
        }

        printf("%d", i);
        fflush(stdout);
        Sleep(1);

        i += 1;
        for (int j = 0; j < 3; j++)
        {
            event_set_state(i_bits[j], (i >> j) & 1);
        }

        if (i >= 5)
        {
            break;
        }
    }
}
